HubSpot V2 doc (#786)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/hubspot-v2.mdx

To associate a `track` call with a user, you need to specify the user's `email` under `context.traits`. Also, RudderStack associates the `track` events with the same user after an `identify` request is successfully made.

A sample `track` event is as shown:

```javascript
rudderanalytics.track(
  "Order Completed", {
    value: 30,
  }, {
    context: {
      traits: {
        firstname: "Alex",
        city: "New Orleans",
        country: "USA",
        phone: "+1-202-555-0146",
        email: "alex@example.com",
      },
    },
  }
)
```

### Property mappings

The following table lists the **optional** property mappings between RudderStack and HubSpot for the `track` call:

| RudderStack property         | HubSpot property      | 
| :--------------------------- | :--------------------- |
| `traits.utk`/`context.traits.utk`/`properties.utk`  | `utk`     | 
| `email`        | `email`       |
| `properties.occurred_at`/`originalTimestamp`  | `occurredAt`       |         
| `traits.objectId`/`context.traits.objectId`/`properties.objectId`    | `objectId`     |            

### Custom behavioral events

> 

Reviewer: To use custom behavorial events, one needs to opt for the private app access token only. And, private app must have permission  : `analytics.behavioral_events.send ` @akashrpo please confirm

Reviewer: Hi, sorry to go to and fro here, `To use custom behavorial events, one needs to opt for the private app access token only.` <-- this information needed to be deleted. We can send custom behavioral events with an API key as well

Reviewer 2: We can use custom behavioral events for both the Authorisation Type.
![image](https://user-images.githubusercontent.com/64877812/182371060-f7904847-d298-4401-ac84-074abeb2a9f6.png)

---
title: "HubSpot V2"

> description: Step-by-step guide on sending your event data from RudderStack to HubSpot using the new API

Reviewer 3: ```suggestion
description: Step-by-step guide on sending your event data from RudderStack to HubSpot using the new HubSpot API.
```

Reviewer 3: As per our discussion, highlighting in the meta itself for clarity and differentiation.

<div class="infoBlock">
To know more about the difference between cloud mode and device mode in RudderStack, refer to the <Link to="/destinations/rudderstack-connection-modes/">RudderStack Connection Modes</Link> guide.
</div>

Once you have confirmed that the source platform supports sending events to HubSpot, follow these steps:

1. From your [RudderStack dashboard](https://app.rudderstack.com/), add the source. Then, from the list of destinations, select **HubSpot**.
2. Assign a name to the destination and click on **Continue**.

## Connection settings

To successfully configure HubSpot as a destination using the **new API**,  you can choose any of the two authorization types explained below:

#### Private Apps authorization type

The **Private Apps** authorization type allows you to use HubSpot's APIs to access specific data from your HubSpot account.

<img src="../../assets/event-stream-destinations/hubspotv2-connection-settings-2.png" alt="HubSpot connection settings" />

- **Access Token**: Enter your HubSpot access token from the **Private Apps** section in the HubSpot dashboard.

<div class="infoBlock">

> Refer to the <Link to="#faq">FAQ</Link> section to know where to get the HubSpot access token.

Reviewer 3: ```suggestion
Refer to the <Link to="#faq">FAQ</Link> section for more information on getting the HubSpot access token.
```

</div>

> - **API Version**: Select **New API** from the dropdown.

Reviewer 3: ```suggestion
- **API Version**: Select **New API (v3)** from the dropdown.
```

#### API Key authorization type

> The **API Key (deprecating soon)** authorization type allows you to create custom applications using HubSpot's APIs.

Reviewer 3: ```suggestion
The **API Key (deprecating soon)** authorization type lets you create custom applications using HubSpot's APIs.
```

<div class="warningBlock">

> HubSpot API keys will be deprecated soon. Refer to <Link to="https://knowledge.hubspot.com/integrations/how-do-i-get-my-hubspot-api-key">HubSpot documentation</Link> for more information.

Reviewer 3: ```suggestion
This authorization type will be deprecated soon. Refer to <Link to="https://knowledge.hubspot.com/integrations/how-do-i-get-my-hubspot-api-key">HubSpot documentation</Link> for more information.
```

</div>

<img src="../../assets/event-stream-destinations/hubspotv2-connection-settings-1.png" alt="HubSpot connection settings" />

- **Hub ID**: Enter your HubSpot Hub ID.
- **API Key**: Enter your HubSpot API key. This field is mandatory to send data to HubSpot via the RudderStack cloud mode.

<div class="infoBlock">
Refer to the <Link to="#faq">FAQ</Link> section to know where to get the Hub ID and API Key.
</div>

> - **API Version**: Select **New API** from the dropdown.

Reviewer 3: ```suggestion
- **API Version**: Select **New API (v3)** from the dropdown.
```

### Event and property names

<img src="../../assets/event-stream-destinations/hubspotv2-connection-settings-3.png" alt="HubSpot connection settings" />

- **HubSpot property name to be used as lookup field**: Enter the unique HubSpot property to be used as a lookup field for <Link to="https://developers.hubspot.com/docs/api/crm/contacts">HubSpot contacts</Link>.

> The below settings are only applicable for the **Private Apps** authorization type:

Reviewer 3: ```suggestion
The below settings are only applicable for the <Link to="#private-apps-authorization-type">**Private Apps**</Link> authorization type:
```

- **RudderStack Event Name**: Enter the RudderStack event name you want to map to the HubSpot custom behavioral event name.
- **HubSpot Event Name**: Enter the HubSpot event name you want to map to the above-mentioned RudderStack event.

> - **Map your Event Property**: Enter the RudderStack and HubSpot event properties you want to map for the above-mentioned custom behavioral events. You can specify multiple properties for a single event.

Reviewer 4: Can we elaborate this point this is not giving full context of what should be done.
You can refer this loom demo video on how we are using this

Reviewer 4: Apologies @ishachopra0401 missed it
https://www.loom.com/share/1eff601602c143dfb53f17a0487a6da6

Reviewer 2: Custom Behavioural Events - 10:40

<div class="infoBlock">

> Refer to the <Link to="https://developers.hubspot.com/docs/api/analytics/events">custom behavioral events</Link> section to know more about them.

Reviewer 3: ```suggestion
Refer to the <Link to="https://developers.hubspot.com/docs/api/analytics/events">Custom Behavioral Events</Link> guide for more information.
```

Reviewer 3: would it be helpful to link to our own section on Custom behavioral events here, instead of redirecting to HubSpot docs?

Reviewer 3: ```suggestion
Refer to the <Link to="#custom-behavioral-events">Custom behavioral events</Link> section below for more information.
```

</div>

### Client-side events filtering

<img src="../../assets/event-stream-destinations/hubspot-connection-settings-3.png" alt="HubSpot connection settings" />

<div class="infoBlock">
These settings are applicable only if you're sending events to HubSpot via the web device mode. Refer to the <Link to="/sources/sdks/event-filtering/">Client-side Event Filtering</Link> guide for more information on this feature.
</div>

## Identify

> RudderStack sends the <Link to="/event-spec/standard-events/identify">`identify`</Link> call to check if a contact exists, if yes, it performs the update through the new update endpoints.

Reviewer 3: ```suggestion
RudderStack sends the <Link to="/event-spec/standard-events/identify">`identify`</Link> call to check if a contact exists in HubSpot. If it does, RudderStack performs the update through the new update endpoints.
```

Reviewer 3: "If it does, RudderStack performs the update through the new update endpoints."

can we simplify this to "If it does, RudderStack updates the user with the new information. Otherwise, it creates a new user." cc @shrouti1507 @akashrpo

Reviewer 2: Looks good

> RudderStack sends the <Link to="/event-spec/standard-events/identify">`identify`</Link> call to check if a contact exists, if yes, it performs the update through the new update endpoints.

Reviewer 3: @ishachopra0401 think we missed this update.

"If yes, RudderStack updates the user with the new information. Otherwise, it creates a new user." -- we need to specify the new user creation bit as well.

A sample `identify` call is as shown:

```javascript
rudderanalytics.identify({
  firstName: "Alex",
  city: "New Orleans",
  country: "USA",
  phone: "+1-202-555-0146",
  email: "alex@example.com",
  custom_flavor: "chocolate",
  custom_date: 1574769933368,
  custom_date1: new Date("2019-10-14T11:15:53.296Z"),
})
```

## Track

A <Link to="/event-spec/standard-events/track/">`track`</Link> call lets you record various user actions and any properties associated with them. 

> To associate a `track` call with a user, you need to specify the user's `email` under `context.traits`. Also, RudderStack associates the `track` events with the same user after an `identify` request is successfully made.

Reviewer 3: @shrouti1507 @akashrpo For HubSpot, should `identify` be mandatorily called before `track` or is it optional? cc @ishachopra0401

Reviewer 2: It is not mandatory however if there is no contact in HS then there won't be any association of events as we won't have any identifier but still the track will go through. https://developers.hubspot.com/docs/api/analytics/events
